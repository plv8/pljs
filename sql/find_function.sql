CREATE FUNCTION callee(a int) RETURNS int AS $$ return a * a $$ LANGUAGE pljs;
CREATE FUNCTION sqlf(int) RETURNS int AS $$ SELECT $1 * $1 $$ LANGUAGE sql;
CREATE FUNCTION caller(a int, t int) RETURNS int AS $$
  var func;
  if (t == 1) {
    func = pljs.find_function("callee");
  } else if (t == 2) {
    func = pljs.find_function("callee(int)");
  } else if (t == 3) {
    func = pljs.find_function("sqlf");
  } else if (t == 4) {
    func = pljs.find_function("callee(int, int)");
  } else if (t == 5) {
    try{
      func = pljs.find_function("caller()");
    }catch(e){
      func = function(a){ return a };
    }
  }
  return func(a);
$$ LANGUAGE pljs;

SELECT caller(10, 1);
SELECT caller(10, 2);
SELECT caller(10, 3);
SELECT caller(10, 4);
SELECT caller(10, 5);


-- test find_function permissions failure
CREATE FUNCTION perm() RETURNS void AS $$ pljs.elog(NOTICE, 'nope'); $$ LANGUAGE pljs;

CREATE ROLE someone_else;

REVOKE EXECUTE ON FUNCTION perm() FROM public;

SET ROLE TO someone_else;
DO $$ const func = pljs.find_function('perm') $$ LANGUAGE pljs;
DO $$ const func = pljs.find_function('perm()') $$ LANGUAGE pljs;

RESET ROLE;
DROP ROLE someone_else;

DROP FUNCTION perm();
