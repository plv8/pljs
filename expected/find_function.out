CREATE FUNCTION callee(a int) RETURNS int AS $$ return a * a $$ LANGUAGE pljs;
CREATE FUNCTION sqlf(int) RETURNS int AS $$ SELECT $1 * $1 $$ LANGUAGE sql;
CREATE FUNCTION caller(a int, t int) RETURNS int AS $$
  var func;
  if (t == 1) {
    func = pljs.find_function("callee");
  } else if (t == 2) {
    func = pljs.find_function("callee(int)");
  } else if (t == 3) {
    func = pljs.find_function("sqlf");
  } else if (t == 4) {
    func = pljs.find_function("callee(int, int)");
  } else if (t == 5) {
    try{
      func = pljs.find_function("caller()");
    }catch(e){
      func = function(a){ return a };
    }
  }
  return func(a);
$$ LANGUAGE pljs;
SELECT caller(10, 1);
 caller 
--------
    100
(1 row)

SELECT caller(10, 2);
 caller 
--------
    100
(1 row)

SELECT caller(10, 3);
ERROR:  execution error
DETAIL:  Error: javascript function is not found for "sqlf"
    at caller (<function>:9:30)

SELECT caller(10, 4);
ERROR:  execution error
DETAIL:  Error: javascript function is not found for "callee(int, int)"
    at caller (<function>:11:30)

SELECT caller(10, 5);
 caller 
--------
     10
(1 row)

-- test find_function permissions failure
CREATE FUNCTION perm() RETURNS void AS $$ pljs.elog(NOTICE, 'nope'); $$ LANGUAGE pljs;
CREATE ROLE someone_else;
REVOKE EXECUTE ON FUNCTION perm() FROM public;
SET ROLE TO someone_else;
DO $$ const func = pljs.find_function('perm') $$ LANGUAGE pljs;
WARNING:  failed to find or no permission for js function perm
DO $$ const func = pljs.find_function('perm()') $$ LANGUAGE pljs;
WARNING:  failed to find or no permission for js function perm()
RESET ROLE;
DROP ROLE someone_else;
DROP FUNCTION perm();
